include/websocket.hpp

```diff
@@ -0,0 +1,472 @@
+#include <boost/asio/buffers_iterator.hpp>
+#include <boost/asio/strand.hpp>
+#include <boost/beast/core.hpp>
+#include <boost/beast/websocket.hpp>
+#include <iostream>
+#include <memory>
+#include <nlohmann/json.hpp>
+#include <queue>
+#include <string>
+#include <thread>
+
+namespace beast = boost::beast;          // from <boost/beast.hpp>
+namespace http = beast::http;            // from <boost/beast/http.hpp>
+namespace websocket = beast::websocket;  // from <boost/beast/websocket.hpp>
+namespace net = boost::asio;             // from <boost/asio.hpp>
+using tcp = boost::asio::ip::tcp;        // from <boost/asio/ip/tcp.hpp>
+using json = nlohmann::json;             // from <nlohmann/json.hpp>
+
+typedef std::function<void(json &)>
+    Callback;  // callback function which takes the json and returns nothing
+
+/// @brief Class to store the request and function given by user
+struct RequestContent {
+  // id will be the id for subscription while id+1 will be id for unsubscription
+  int id_;
+  // subscription and unsubscription method
+  std::string subscribe_method_;
+  std::string unsubscribe_method_;
+
+  // callback to call in case of notification
+  Callback cb_;
+  // params to be passed with the subscription string
+  json params_;
+
+  /// @brief constructor
+  RequestContent(int id, std::string subscribe_method,
+                 std::string unsubscribe_method, Callback cb, json params)
+      : id_(id), subscribe_method_(subscribe_method),
+        unsubscribe_method_(unsubscribe_method), cb_(cb), params_(params) {}
+
+  /// @brief Get the json request to do subscription
+  /// @return the json that can be used to make subscription
+  json get_subscription_request() {
+    json req = {{"jsonrpc", "2.0"},
+                {"id", id_},
+                {"method", subscribe_method_},
+                {"params", params_}};
+    return req;
+  }
+
+  /// @brief Json data to write to unsubscribe
+  /// @param subscription_id the id to subscribe for
+  /// @return request json for unsubscription
+  json get_unsubscription_request(int subscription_id) {
+    json params = {subscription_id};
```

Comment:
Implementation for all the methods should be moved to a seperate c++ file. hpp file should only contain the defination of the class.

Reply:
Yeah I am planning to do this but as I have already mentioned that the tests run standalone but don't when integrated, so for debugging purposes I am trying to have as less change as possible.
Will move it to a cpp file once I am able to debug why the tests cause seg fault,